### PR TITLE
Orchestration

### PR DESCRIPTION
**Pull request by ViktorKaz**
I implemented the orchestration and evaluation code with the new three loop orchestrator design. Let me know if you have comments.

@jasonlines can you please check line 58 in sktime.utils.results_writing.write_results_to_uea_format function. If the probabilities are not passed the function seems to save everything on a single line. I think it is just a matter of unindenting line 58 so that a new line is written irrespective of whether propas were provided or not but I have not experimented with it.

### CODE REVIEW

File: sktime/experiments/analysis.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+    Parameters
+    ----------
+    result: sktime result object
+        class for storing the results
+    """
+
+    def __init__(self, 
+                 results):
+
+        self._results_list = results.load()
```
Comment (mloning) [inline sktime/experiments/analysis.py]: instead of calling `load` here, we could call it in line 47 when you iterate over results
```suggestion
        self._results_list = results
```

File: sktime/experiments/analysis.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        metric: `sktime.analyse_results.scores`
+            Error function 
+        Returns
+        --------
+        pickle of pandas DataFrame
+            ``estimator_avg_error`` represents the average error and standard deviation achieved by each estimator. ``estimator_avg_error_per_dataset`` represents the average error and standard deviation achieved by each estimator on each dataset.
+        """
+        #load all predictions
+        losses = Losses(metric)
+        for res in self._results_list:
```
Comment (mloning) [inline sktime/experiments/analysis.py]:
```suggestion
        for res in self._results_list.load():
```
The `load` method would have to be an iterator then.

File: sktime/experiments/analysis.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+            average = np.average(scores_dict[k])
+            n = len(scores_dict[k])
+            std_error = np.std(scores_dict[k])/np.sqrt(n)
+            result[k]=[average,std_error]
+        
+        res_df = pd.DataFrame.from_dict(result, orient='index')
+        res_df.columns=['avg_score','std_error']
+        res_df = res_df.sort_values(['avg_score','std_error'], ascending=[1,1])
+
+        return res_df.round(3)
```
Comment (mloning) [inline sktime/experiments/analysis.py]: I wouldn't round results internally, and leave it to the user to decide whether or not to round the returned output
```suggestion
        return res_df
```
Same applies to other methods

File: sktime/model_selection.py [hunk 4]
```diff
@@ -225,3 +233,35 @@ def __init__(self, estimator, param_grid, scoring=None, fit_params=None,
             # using mean squared error as default for regressors
             elif isinstance(self.estimator, BaseRegressor):
                 self.scoring = make_scorer(mean_squared_error)
+
+class SKTimeCV(ABC):
```
Comment (mloning) [inline sktime/model_selection.py]: Do we still need this class?

File: sktime/model_selection.py [hunk 5]
```diff
@@ -225,3 +233,35 @@ def __init__(self, estimator, param_grid, scoring=None, fit_params=None,
...
+    def split(self, data):
+        """
+        Implements the splits to return the actual data instead of the 
+        """
+
+class PredefinedSplit:
```
Comment (mloning) [inline sktime/model_selection.py]: PredefinedSplit is already a class in sklearn, perhaps call it PresplitFilesCV or so to avoid ambiguity

File: sktime/highlevel.py [hunk 6]
```diff
@@ -127,21 +126,26 @@ def pred_horizon(self):
...
     perform a pescribed task
-
     Parameters
     ----------
     estimator : BaseEstimator
         An instance of an appropriately initialized
         low-level estimator
     """
-    def __init__(self, estimator):
+    def __init__(self, estimator, *args, **kwargs):
```
Comment (mloning) [inline sktime/highlevel.py]: I would make it an explicit kwarg with defaults. 
```suggestion
    def __init__(self, estimator, name=None):
```

File: sktime/highlevel.py [hunk 7]
```diff
@@ -127,21 +126,26 @@ def pred_horizon(self):
...
         # construct and initialize the estimator
         self._estimator = estimator
         self._case = None
         self._task = None
-        # self._traits = {"tags": None}  # traits for matching strategies with tasks
+        self._name = None
```
Comment (mloning) [inline sktime/highlevel.py]:
```suggestion
        self._name = name if name is not None else self.__class__.__name__
```

File: sktime/experiments/orchestrator.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                    if predict_on_runtime:
+                        y_pred = strategy.predict(dts_loaded.iloc[test])
+                        y_true = dts_loaded[task.target].iloc[test].values
+                        self._result.save(dataset_name=data.dataset_name, 
+                                          strategy_name=strategy.name, 
+                                          y_true=y_true.tolist(), 
+                                          y_pred=y_pred.tolist(),
+                                          cv_fold=cv_fold)
+                    if save_strategies:
+                        strategy.save(dataset_name=data.dataset_name,
```
Comment (mloning) [inline sktime/experiments/orchestrator.py]: I think the strategy `save` method should not take the dataset and cv as input args, instead the `self._results` object should have a `save_fitted_strategy` class which determines the correct path and then calls the strategy's `save` method.

File: sktime/experiments/orchestrator.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        
+        for task, data in zip(self._tasks, self._datasets):
+            dts_loaded = data.load()
+            for strategy in self._strategies:
+                for cv_fold, (train, test) in enumerate(self._cv.split(dts_loaded)):
+                    
+                    strategy.fit(task, dts_loaded.iloc[train])
+
+                    if predict_on_runtime:
```
Comment (mloning) [inline sktime/experiments/orchestrator.py]: Optionally returning predictions on the training set may be useful in some cases too

File: sktime/experiments/analysis.py (hunk 1 above)
Comment (ViktorKaz) [reply sktime/experiments/analysis.py]: looks kind of similar to me, is there a performance or other gain?

File: sktime/experiments/analysis.py (hunk 3 above)
Comment (ViktorKaz) [reply sktime/experiments/analysis.py]: ok, will push shortly, removing the rounding.

File: sktime/model_selection.py (hunk 4 above)
Comment (ViktorKaz) [reply sktime/model_selection.py]: SKTimeCV? Yes, I believe so, PredefinedSplit and other custom CV objects should inherit from it.

File: sktime/model_selection.py (hunk 5 above)
Comment (ViktorKaz) [reply sktime/model_selection.py]: ok

File: sktime/highlevel.py (hunk 6 above)
Comment (ViktorKaz) [reply sktime/highlevel.py]: ok, *args, **kwargs was in the original code. I assumed this was needed for future extensions.

File: sktime/experiments/orchestrator.py (hunk 8 above)
Comment (ViktorKaz) [reply sktime/experiments/orchestrator.py]: not sure how this will work. You need to provide the path when saving the strategy. What you can do is instead of providing the different arguments that form the path separately you can create the path string inside the results.save_fitted_strategy() method and than call strategy.save(path). However, I don't think that it will make the code more readable.

File: sktime/experiments/orchestrator.py (hunk 9 above)
Comment (ViktorKaz) [reply sktime/experiments/orchestrator.py]: sure, let's discuss what the requirements are and I can make this change.

File: sktime/experiments/analysis.py (hunk 1 above)
Comment (mloning) [reply sktime/experiments/analysis.py]: You wouldn't load all of the results (true and predicted labels) into memory at the same time

File: sktime/experiments/orchestrator.py (hunk 8 above)
Comment (mloning) [reply sktime/experiments/orchestrator.py]: Yes, exactly, that's what I mean

File: sktime/model_selection.py (hunk 4 above)
Comment (mloning) [reply sktime/model_selection.py]: Instead of creating new base classes for cv iterators, I'd rather integrate our objects into the existing sklearn hierarchy of CV classes. (Also currently the inheritance does not do anything unless I'm not seeing something - if the split method should be abstract, it needs the @abstractmethod decorator, in either case, we initially agreed to avoid abstract methods for now if I remember correctly)

Comment (ViktorKaz) [reply sktime/model_selection.py]: I forgot the decorator in the the abstract class, this is a mistake.

I am thinking along the lines of the extension guidelines, i.e. if a user wants to write a new CV object they would inherit from SKTimeCV and implement spilt(). At this time, the code is fairly simple so might be an overkill but this can be a future-proof solution.

How do you propose to integrate the cv classes in the sklearn hierachy, by inheriting from BaseCrossValidator? I don't think that this will work particularly well in our case as the split() method that is inheritted will not be appropriate for our predefined splits use case.

In any case, let's delete the abstract class, it doesn't do anything at the moment and there is no scope to write more CV methods in the near future.

File: sktime/experiments/analysis.py (hunk 1 above)
Comment (ViktorKaz) [reply sktime/experiments/analysis.py]: Your suggestion also loads everything into memory at the same time it just does it in a different place. 

The memory requirements for loading all the results should not be very big, I would assume.

Comment (mloning) [reply sktime/experiments/analysis.py]: Not if the `load` method is an iterator ...
